cmake/OpenCVDetectCUDA.cmake

```diff
@@ -70,29 +76,52 @@ if(CUDA_FOUND)
     unset(CUDA_ARCH_PTX CACHE)
   endif()
 
+  macro(ocv_filter_available_architecture result_list TARGET_ARCH_LIST)
+    string(REPLACE " " ";" CC_LIST ${TARGET_ARCH_LIST})
+    foreach(target_arch IN LISTS CC_LIST)
+      string(REPLACE "." "" target_arch_short ${target_arch})
+      set(NVCC_OPTION "-gencode;arch=compute_${target_arch_short},code=sm_${target_arch_short}")
+      execute_process( COMMAND "${CUDA_NVCC_EXECUTABLE}" ${NVCC_OPTION} "${OpenCV_SOURCE_DIR}/cmake/checks/OpenCVDetectCudaArch.cu"
```

Comment:
It would be nice to ensure that its result is cached (or upstream variable is cached), so `cmake .` call doesn't trigger that again in a loop.